ManaBarUI and ManaSystem Updates

MatthewMorisky opened the pull request:
Adjusted the code to make it run more efficiently. Also adjusted the mana using a lerp function when mana has changed instead of the old method of immediate change. Lastly I added error checking which will display an error to the Debug.Log when GainMana() or ConsumeMana() receive negative values. Adjusted comments accordingly.

Closes #82

MatthewMorisky committed:
ManaBarUI and ManaSystem Updates

Adjusted the code to make it run more efficiently. Also adjusted the mana using a lerp function when mana has changed instead of the old method of immediate change. Lastly I added error checking which will display an error to the Debug.Log when GainMana() or ConsumeMana() receive negative values. Adjusted comments accordingly.

Kainkun committed:
Merge branch 'main' into ManaSystemAndUI

Kainkun commented on Assets/Scenes/ManaScene.unity:
```diff
@@ -523,6 +523,8 @@ MonoBehaviour:
   m_EditorClassIdentifier: 
   ManaBarFill: {fileID: 1293296639}
+  adjustMana: 0
+  lerpDuration: 1.5
```
a bit long, maybe make it 0.5 seconds snappier visuals feel better especially for important information

Kainkun commented on Assets/Scripts/ManaBarUI.cs:
```diff
@@ -1,20 +1,29 @@
 using UnityEngine;
 using UnityEngine.UI;
 
 public class ManaBarUI : MonoBehaviour
```
I just realized this can be made more generic and be used for any bar like health bars. change the class name and the file name (they have to be the same) and change all the variables to not be mana related

Kainkun commented on Assets/Scripts/ManaSystem.cs:
```diff
@@ -29,62 +28,89 @@ public float GetCurrentMana()
-            currentMana = maxMana;
+            float manaBefore = currentMana;
+            currentMana += manaGain;
+            Mathf.Clamp(currentMana, 0, maxMana);
```
Mathf.Clamp does not take variables by reference. instead, it returns the new value so you'll have to do current mana = Mathf.Clamp etc.

Kainkun commented on Assets/Scripts/ManaBarUI.cs:
```diff
@@ -1,20 +1,29 @@
 using System.Collections.Generic;
 using UnityEngine;
 using UnityEngine.UI;
 
```
i added [EasyButtons](https://github.com/madsbangh/EasyButtons) to this project to make it easier to test code. If you include `using EasyButtons` and add the attribute `[Button]` right above each function, you'll be able to run them by clicking a button in the inspector. You can also limit it so that the buttons can't be clicked during edit mode with `[Button(mode = buttonmodes.EnabledInPlayMode)]` or something close to that i forgot let me know if you need help

Kainkun reviewed (changes requested):
I added some suggestions to your code. The main issue is that the clamp isn't used correctly.
Also, not sure if this is an issue but I think two of the same coroutines can be run at the same time btw.

Kainkun commented:
Oh! also, instead of having the Empty ManaBarUI gameobject, you can directly add the ManaBarUI component/script to the ManaBarFill itself. That will keep things organized. And then you would also be able to use GetComponent<ManaBarUI>() in Start() to get a reference to the UI element without having to click and drag the object into the script.

MatthewMorisky committed:
Merge branch 'main' into ManaSystemAndUI

MatthewMorisky committed:
Merge branch 'ManaSystemAndUI' of https://github.com/AztecGameLab/LongTermGame5 into ManaSystemAndUI

MatthewMorisky committed:
Merge branch 'main' into ManaSystemAndUI

MatthewMorisky committed:
Implemented all of Kain's suggestions

Renamed all objects to a generic Fill Bar title to separate them from mana. Adjusted the clamp function so it actually has a purpose. Implemented buttons so the code can be easily tested. Further augmented comments to suit these changes.

Kainkun committed:
Merge branch 'main' into ManaSystemAndUI

Kainkun commented on Assets/Scripts/FillBarUI.cs:
```diff
@@ -0,0 +1,34 @@
+
+    void Start()
+    {
+        BarFill = gameObject.GetComponent(typeof(Image)) as Image;
```
This works perfectly fine but get component is usually used more like this `BarFill = gameObject.GetComponent<Image>();`

Kainkun commented on Assets/Scripts/FillSystem.cs:
```diff
@@ -0,0 +1,145 @@
+     */
+    void Start()
+    {
+        FillBarUI = gameObject.GetComponent(typeof(FillBarUI)) as FillBarUI;
```
same as previous comment. it can be FillBarUI = gameObject.GetComponent<FillBarUI>();

Kainkun commented on Assets/Scripts/FillSystem.cs:
```diff
@@ -0,0 +1,145 @@
+        if (valueGained < 0)
+        {
+            Debug.LogError("The Gain() function expects a positive valueGained not a negative one");
+            Debug.LogError("Did you intend to use the Consume() function?");
```
might be better to use a new line char '\n' instead of printing twice. It looks like two errors when its technically one error.  That's not important though just wanted to mention the idea

Kainkun commented on Assets/Scripts/FillSystem.cs:
```diff
@@ -0,0 +1,145 @@
+            Debug.LogError("Did you intend to use the Gain() function?");
+            return false;
+        }
+        else if (currentFill > cost)
```
shouldn't this be `>=` ?

Kainkun commented on Assets/Scripts/FillSystem.cs:
```diff
@@ -0,0 +1,145 @@
+        {
+            currentFill = 0;
+            StartCoroutine(FillBarUI.ChangeFillAmount(fillBefore, currentFill, maxFill));
+            return true;
```
I think this would return true even if you have no mana? I think there needs to be a currentFill <= 0 to return false if you try to use mana when you have 0 mana because that shouldn't return true even if it "canBeReducedPastZero" lmk if that doesn't make sense

MatthewMorisky replied:
The reason I implemented it this way is a character might take damage past zero. For mana using abilities the "canBeReducedPastZero" flag should always be false.

Kainkun reviewed (approved):
looks good overall. I added some comments on what you might want to tweak. I think the last comment is the most important. I'll accept this PR since it's mostly working, but feel free to send me another PR if you decide to make some tweaks.
Thanks!
